apps/api/src/http/internal/rest/webhooks/stripe/checkout_session.rs

```diff
@@ -0,0 +1,253 @@
+};
+use shared::database::product::SubscriptionProductId;
+use shared::database::queries::{filter, update};
+use shared::database::user::UserId;
+
+use crate::global::Global;
+use crate::http::error::ApiError;
+use crate::http::v3::rest::egvault::redeem::grant_entitlements;
+use crate::transactions::{TransactionError, TransactionResult, TransactionSession};
+
+pub async fn completed(
+	global: &Arc<Global>,
+	mut tx: TransactionSession<'_, ApiError>,
+	session: stripe::CheckoutSession,
+) -> TransactionResult<Option<SubscriptionId>, ApiError> {
+	let Some(metadata) = session.metadata else {
+		// ignore metadata sessions that don't have metadata
+		return Ok(None);
+	};
+
+	if session.mode == stripe::CheckoutSessionMode::Setup {
+		// setup session
+		// the customer successfully setup a new payment method, now set it as the
+		// default payment method
+
+		let setup_intent = session
+			.setup_intent
+			.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?
+			.id();
+		let setup_intent = stripe::SetupIntent::retrieve(&global.stripe_client, &setup_intent, &[])
+			.await
+			.map_err(|e| {
+				tracing::error!(error = %e, "failed to retrieve setup intent");
+				TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+			})?;
+
+		let customer_id = session.customer.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?.id();
+		let Some(payment_method) = setup_intent.payment_method.map(|p| p.id()) else {
+			return Ok(None);
+		};
+
+		let customer = stripe::Customer::update(
+			&global.stripe_client,
+			&customer_id,
+			stripe::UpdateCustomer {
+				invoice_settings: Some(stripe::CustomerInvoiceSettings {
+					default_payment_method: Some(payment_method.to_string()),
+					..Default::default()
+				}),
+				..Default::default()
+			},
+		)
+		.await
+		.map_err(|e| {
+			tracing::error!(error = %e, "failed to update customer");
+			TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+		})?;
+
+		if let Some(user_id) = customer
+			.metadata
+			.and_then(|m| m.get("USER_ID").and_then(|i| UserId::from_str(i).ok()))
+		{
+			if let Some(ProviderSubscriptionId::Stripe(sub_id)) = tx
+				.find_one(
+					filter::filter! {
+						SubscriptionPeriod {
+							#[query(flatten)]
+							subscription_id: SubscriptionId {
+								user_id,
+							},
+							#[query(selector = "lt")]
+							start: chrono::Utc::now(),
+							#[query(selector = "gt")]
+							end: chrono::Utc::now(),
+						}
+					},
+					None,
+				)
+				.await?
+				.and_then(|p| p.provider_id)
+			{
+				stripe::Subscription::update(
+					&global.stripe_client,
+					&sub_id,
+					stripe::UpdateSubscription {
+						default_payment_method: Some(&payment_method),
+						..Default::default()
+					},
+				)
+				.await
+				.map_err(|e| {
+					tracing::error!(error = %e, "failed to update subscription");
+					TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+				})?;
+			}
+		}
```

Comment:
Is this basically jsut changing the payment method when they pay with a different payment method? Is this desirable? 
What happens if sub period is gifted?

Reply:
No, it's changing the payment method when they want to change it using the button on the website